calico.py

```diff
@@ -441,14 +449,87 @@ def _call_ipam_plugin(self, env):
+                self.network_config["ipam"]["subnet"] = str(pod_cidr)
+
+            # Call the IPAM plugin.
+            _log.debug("Calling host-local IPAM plugin")
+            code, response = self._call_binary_ipam_plugin(env)
         else:
+            # Using some other IPAM plugin - call it.
             _log.debug("Using binary plugin")
             code, response = self._call_binary_ipam_plugin(env)
 
         # Return the IPAM return code and output.
         _log.debug("IPAM response (rc=%s): %s", code, response)
         return code, response
 
+    def _get_kubernetes_pod_cidr(self):
+        """
+        Attempt to get the Kubernetes pod CIDR for this node.
+        First check if we've written it to disk.  If so, use that value.  If
+        not, then query the Kubernetes API for it.
+        """
+        local_store_path = os.path.abspath("podCidr.cache")
+        if os.path.exists(local_store_path):
+            _log.info("Loading podCidr from disk: %s", local_store_path)
+            with open(local_store_path, "r") as f:
+                pod_cidr = f.read().strip()
+                _log.debug("Loaded podCidr from disk: %s", pod_cidr)
+        else:
+            _log.info("Getting node.spec.podCidr from API, kubeconfig: %s",
+                      self.kubeconfig_path)
+            if not self.kubeconfig_path:
+                # Fow now, kubeconfig is the only supported auth method.
+                print_cni_error(ERR_CODE_GENERIC, "Missing kubeconfig",
+                        "usePodCidr requires specification of kubeconfig file")
+                sys.exit(ERR_CODE_GENERIC)
+
+            # Query the API for this node.  Default node name to the hostname.
+            try:
+                api = HTTPClient(KubeConfig.from_file(self.kubeconfig_path))
+                node = None
+                for n in Node.objects(api):
+                    if n.obj["metadata"]["name"] == self.k8s_node_name:
+                        _log.debug("Checking node: %s", n.obj["metadata"]["name"])
+                        node = n
+                        break
+                if not node:
+                    raise KeyError("Unable to find node in API: %s", self.k8s_node_name)
+                _log.debug("Found node %s: %s: ", node.obj["metadata"]["name"],
+                           node.obj["spec"])
+            except Exception:
+                print_cni_error(ERR_CODE_GENERIC, "Error querying Kubernetes API",
+                        "Failed to get podCidr from Kubernetes API")
+                sys.exit(ERR_CODE_GENERIC)
+            else:
+                pod_cidr = node.obj["spec"].get("podCIDR")
+                if not pod_cidr:
+                    print_cni_error(ERR_CODE_GENERIC, "Missing podCidr",
+                            "No podCidr for node %s" % self.k8s_node_name)
+                    sys.exit(ERR_CODE_GENERIC)
+                else:
+                    # Success - write to disk so we don't have to
```

Comment:
This is really scary. What assurances can you give me that the IP range will never change?